vcf.c

```diff
@@ -4261,6 +4261,30 @@ int bcf_get_variant_type(bcf1_t *rec, int ith_allele)
     if ( rec->d.var_type==-1 ) bcf_set_variant_types(rec);
     return rec->d.var[ith_allele].type;
 }
+inline static int _has_variant_type(int type, int bitmask, enum bcf_variant_match mode)
+{
+    if ( mode==overlap ) return type & bitmask;
+
+    // VCF_INDEL is always set with VCF_INS and VCF_DEL by bcf_set_variant_type[s], but the bitmask may
+    // ask for say `VCF_INS` or `VCF_INDEL` only
+    if ( bitmask&(VCF_INS|VCF_DEL) && !(bitmask&VCF_INDEL) ) type &= ~VCF_INDEL;
+    else if ( bitmask&VCF_INDEL && !(bitmask&(VCF_INS|VCF_DEL)) ) type &= ~(VCF_INS|VCF_DEL);
+
+    if ( mode==subset )
+    {
+        if ( ~bitmask & type ) return 0;
+        else return bitmask & type;
+    }
+    return type==bitmask ? type : 0;
+}
```

Comment:
The `bcf_has_variant_type()` interface has one advantage, which is that it's less likely to be caught out by future changes which add more variant types.  Defining a replacement `bcf_get_variant_type2()` in exactly the same way as the original would hit the problem that we have with the old one - you can't easily extend it.  Adding a bitmask parameter would help with this, as callers would be able to set an expectation as to the types returned.

Whatever interface we end up with, it would also be useful to optionally return the `d.var[].n` or `d.n_var` values, as currently they can only be reached by accessing the data structure directly.  This could be done by adding an `int *` parameter (which could be NULL if you're not interested in the value).

Another issue that would be worth considering is that neither the old or proposed interfaces can return an error code.  It would be useful to be able to do that so we can eventually propagate errors up from `bcf_set_variant_types()` which attempts to allocate memory, which could fail.

Reply:
The purpose is explained in the comment. Please re-read, I really don't know how to explain this better :-(

Reply:
The comment explains what it does, but not why it does it.

I also disagree that this is necessary in order to protect against future changes.  We don't have a similar API for BAM flags for example.  Instead it's documented that it is a clear *bit-field* and not something you should be doing naive equality checks against.

The correct way of handling this is via better documentation of the structure and API, instead of inventing a completely new API to do "and" and "or" checks when the language has them built in anyway.

Reply:
It says "why". The user may want to ask for "is this an indel" or "is this an insertion". The rest is just following simple logic. This would of course not be required if you did not insist on non-breaking change. But if VCF_INDEL has to be preserved as a specific value rather than a combination of VCF_INS|VCF_DEL bits, this is needed.

Reply:
Ok I understand.

I'm not saying your solution is wrong, but the usual way I see of doing this is to return the bit-field as-is and then have macros or static inline functions for interpreting it in a friendly fashion.  Eg see the [inode(7)](https://man7.org/linux/man-pages/man7/inode.7.html) man page.

The `stat.st_mode` field can be manually checked with AND/OR ops (eg `if ((sb.st_mode & S_IFMT) == S_IFBLK) {...}`, but there are also function-a-likes such as `if (S_IFBLK(st_mode)) {...}` to do the same thing.  If we wanted something compound, such as either character device or block device, then the logic could be `(sb_st.mode & (S_IFBLK|S_IFCHR)) == (S_IFBLK|S_IFCHR)` which can be made into e.g. `S_IFDEV()` function.

The analogue here would be a `VCF_IS_INS()`, `VCF_IS_DEL()` and `VCF_IS_INDEL()` functions which take the type field and apply appropriate boolean logic to return true or false.  That's sort of what I was expecting mainly due to familiarity with POSIX interfaces, but I can see combined the match type (overlap, subset, exact) and bitmask into a single API achieves the same goal.  So as I say it's neither right nor wrong, but something to consider.    It's possible the POSIX style may be more in keeping with existing htslib APIs (eg `bcf_gt_is_missing`, `bcf_gt_is_phased`, `bcf_float_is_vector_end`, `bam_is_rev` etc).

Reply:
I am happy for this to be changed to whatever you think will be most convenient